README.md
# saltbot
A WIP Discord bot written with discord.py

### Features
- A manager that starts the bot, allowing for starting and stopping the bot within discord.
- Five cogs: bangs, developer, easter, info and util, two of which currently do nothing.
- Ping fuctions for the bot and the manager.
- Bangs command from Miso bot
- Commands about the bot's github and license
- Released under the GNU AGPL v3.0.
- A README and a changelog.

### Requirements
(Note: The versions given have been tested with the bot. It may work with other versions.)
- UNIX or UNIX-like system*
- Python 3.8.x
- python-dotenv 0.14.0
- discord.py 1.4.0
- aiohttp 3.6.2

### Inviting
The bot will become public in the near future. Its invite link will be here!

### Hosting
This program uses python-dotenv to hide the auth token of the bot. If you wish to self-host, you must create a file called ".env" in the main directory. Paste the following into it, where `<token>` is your bot's auth token.
```
TOKEN="<token>"
```
Run botd.py, and from discord, use the command `&botd start`. To stop the bot, use `&stop` and to stop the manager, use `&botd stop`.

### Issues
Please use GitHub's issue tracker to report any issues and to suggest content!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.0.3] - 2020-08-09
### Added
- Duckduckgo bang feature invoked with `&!<bang> <search> [terms]`.
    - Bangs cog, associated with Duckduckgo bang feature
    - Learn more about Duckduckgo bangs at [duckduckgo.com/bangs](https://duckduckgo.com/bangs)
    - Feature ported from [Miso bot](https://github.com/joinemm/miso-bot) under the [MIT License](https://mit-license.org/).
- New requirement aiohttp 3.6.2

### Changed
- Fixed minor spelling errors
- Changed license and github commands slightly
- Updated README
- Updated CHANGELOG
  - Capitalized filename



## [0.0.2] - 2020-08-08
### Added
- Bot reacts with 👋 to messages that start with words like "hello" and "sup".
- Bot responds to "stfu" with "no u".
- Easter cog, currently does nothing and isn't loaded by the main script.
- Info cog.
- Github command in info cog
- License command with pretty embed in info cog
- Echo command in util cog
### Changed
- Bot status 🤭
- Updated README
- Updated changelog
### Removed
- Base cog, along with 0 features

## [0.0.1] - 2020-08-07 - Initial Release
### Features
- A manager that starts the bot, allowing for starting and stopping the bot within discord.
- Ping fuctions for the bot and the manager.
- Three cogs (base, developer and util) that currently do nothing.
- Released under the GNU AGPL v3.0.
- A README and a changelog.

### Requirements
(Note: The versions given have been tested with the bot. It may work with other versions.)
- UNIX or UNIX-like system
- Python 3.8.x
- python-dotenv 0.14.0
- discord.py 1.4.0


botd.py
#!/usr/bin/env python3
import dotenv
import os
import sys
dotenv.load_dotenv(verbose=True)
TOKEN = os.getenv("TOKEN")

import time

import discord
from discord.ext import commands

#----------------------Header------------------------
client = commands.Bot(command_prefix = "&botd ")
client.remove_command('help')
#----------------------Commands----------------------
@client.command()
async def ping(ctx):
    """Ping the bot daemon"""
    await ctx.send(f"*The manager is online. **{round(client.latency * 1000)} ms**.*")
#----------------------Managing main.py--------------
@client.command()
@commands.is_owner()
async def start(ctx):
    """Start main.py"""
    try:
        print("Starting bot...")
        await ctx.send("*Starting bot...*")
        time.sleep(1.23)
        await ctx.send("*Bot started.*")
        os.system("python3 main.py")
        print("Bot stopped.")
        await ctx.send("*Bot stopped.*")
    except:
        print("Error starting main.py")
        await ctx.send("*Unable to start bot.*")
#----------------------Stopping---------------------
@client.command()
@commands.is_owner()
async def stop(ctx):
    """Stop the manager"""
    await ctx.send("*Stopping manager.*")
    sys.exit()
#----------------------Start-------------------------
@client.event
async def on_connect():
    print("Manager connected. Readying...")
@client.event
async def on_ready():
    print("Changing status...")
    await client.change_presence(status=discord.Status.idle)
    print(f"Manager online. Logged in as {client.user.name}")
@client.event
async def on_disconnect():
    print("Manager disconnected.\a")
#----------------------Footer------------------------
client.run(TOKEN)


cogs/info.py
import discord
from discord.ext import commands

class InfoCog(commands.Cog):
    def __init__(self, client):
        self.client = client

    @commands.command()
    async def github(self, ctx):
        await ctx.send("https://github.com/saltedcoffii/saltbot")
        await ctx.send("Have fun and send patches!!")
        print(f"{ctx.message.author} requested the bot's github.")

    @commands.command()
    async def license(self, ctx):
        embed=discord.Embed(title="GNU Affero General Public License v3.0", description="Permissions of this strongest copyleft license are conditioned on making available complete source code of licensed works and modifications, which include larger works using a licensed work, under the same license. Copyright and license notices must be preserved. Contributors provide an express grant of patent rights. When a modified version is used to provide a service over a network, the complete source code of the modified version must be made available.", color=0xffffff)
        embed.set_thumbnail(url="https://www.clipartmax.com/png/middle/231-2314001_legal-center-or-law-advocate-icon-with-symbol-of-justice-weighing-scale.png")
        embed.add_field(name="Permissions", value="✅ Commercial\n✅ Modification\n✅ Distribution\n✅ Patent Use\n✅ Private Use", inline=True)
        embed.add_field(name="Limitations", value="❌ Liability\n❌ Warranty", inline=True)
        embed.add_field(name="Conditions", value="ℹ License and copyright notice\nℹ State changes\nℹ Disclose source\nℹ Network use is distrobution\nℹ Same license", inline=True)
        embed.set_footer(text="This is not legal advice. The full license may be found within the bot's source code.")
        await ctx.send(embed=embed)

def setup(client):
    client.add_cog(InfoCog(client))


cogs/bangs.py
import aiohttp
from discord.ext import commands
import time

class Bangs(commands.Cog):
    def __init__(self, client):
        self.client = client

    @commands.Cog.listener()
    async def on_command_error(self, ctx, error):
        error = getattr(error, "original", error)
        if isinstance(error, commands.CommandNotFound):
            if ctx.message.content.startswith(f"{ctx.prefix}!"):
                ctx.timer = time.time()
                ctx.iscallback = True
                ctx.command = self.client.get_command("!")
                await ctx.command.callback(self, ctx)

    async def resolve_bang(self, ctx, bang, args):
        async with aiohttp.ClientSession() as session:
            params = {"q": "!" + bang + " " + args, "format": "json", "no_redirect": 1}
            url = "https://api.duckduckgo.com"
            async with session.get(url, params=params) as response:
                data = await response.json(content_type=None)
                location = data.get("Redirect")
                if location == "":
                    return await ctx.send(":warning: Unknown bang or found nothing!")

                while location:
                    response = await session.get(location)
                    location = response.headers.get("location")

                await ctx.send(response.url)

    @commands.command(name="!")
    async def bang(self, ctx):
        if not hasattr(ctx, "iscallback"):
            return await ctx.send_help(ctx.command)

        await ctx.trigger_typing()
        try:
            bang, args = ctx.message.content[len(ctx.prefix) + 1 :].split(" ", 1)
            if len(bang) != 0:
                await self.resolve_bang(ctx, bang, args)
        except ValueError:
            await ctx.send("Please search something! If you aren't sure how to use duckduckgo bangs, check out this link: https://duckduckgo.com/bangs")

def setup(client):
    client.add_cog(Bangs(client))


main.py
#!/usr/bin/env python3
import dotenv
import os
import sys
dotenv.load_dotenv(verbose=True)
TOKEN = os.getenv("TOKEN")

import time

import discord
from discord.ext import commands
from discord.ext.commands import has_permissions
print(f"Discord.py Version: {discord.__version__}")

#----------------------Header------------------------
client = commands.Bot(command_prefix = '&')
client.remove_command('help')
#----------------------Commands----------------------
# The commands in this section should, at some point, be moved to the appropriate cog.
@client.command()
async def ping(ctx):
    """Ping the bot"""
    await ctx.send(f"Pong :zany_face: **{round(client.latency * 1000)} ms**")
@client.event
async def on_message(message):
    """Responds to "stfu" with "no u" """
    if message.content.startswith("stfu") and message.content.endswith("stfu"):
        await message.channel.send("no u")
    await client.process_commands(message)
@client.event
async def on_message(message):
    """Reacts with a wave emote to hi or hello and stuff"""
    if message.content.startswith(("Hey ", "Hello ", "Hi ", "Sup ", "Oi ", "yo ", "hey ", "hello ", "hi ", "sup ", "oi ", "yo ")):
        await message.add_reaction("👋")
    await client.process_commands(message)
#----------------------Dying-------------------------
@client.command()
@commands.is_owner()
async def stop(ctx):
    """Stop the bot and start the manager"""
    try:
        await ctx.send("Informing manager...")
        print("Stopping bot...")
        time.sleep(1.23)
        await ctx.send("Recieved SIGTERM, dying.")
        print("Changing status...")
        await client.change_presence(status=discord.Status.idle) # Back to idle for manager. Move to manager script?
        sys.exit()
    except SystemExit:
        sys.exit()
    except:
        print("Unable to stop bot.")
        await ctx.send("Unable to stop bot.")
#----------------------Cogs--------------------------
extensions = [
    "bangs", # provides access to duckduckgo bangs, from miso bot
    "developer", # non functioning atm, intended for developer only commands
    # "easter", # adds some easter eggs, although currently does nothing
    "info", # adds functions like "github" or "liscence"
    "util" # intended for random utility commands
]
if __name__ == "__main__":
    print("\nLoading cogs:")
    for extension in extensions:
        #try:
        client.load_extension(f"cogs.{extension}")
        print(f"\tLoaded {extension}")
        #except:
        #    print(f"\tError loading {extension}")
    print("")
#----------------------Start-------------------------
@client.event
async def on_connect():
    print("Connected. Readying...")
@client.event
async def on_ready():
    print("Changing status...")
    time.sleep(0.5)
    await client.change_presence(status=discord.Status.online, activity=discord.Activity(type=discord.ActivityType.watching, name="gay porn"))
    print(f"Ready. Logged in as {client.user.name}")
@client.event
async def on_disconnect():
    print("Disconnected.\a")
#----------------------Footer------------------------
client.run(TOKEN)
